packages/profile-sync-controller/src/controllers/user-storage/account-syncing/controller-integration.ts

```diff
@@ -0,0 +1,319 @@
+import { isEvmAccountType } from '@metamask/keyring-api';
+import type { InternalAccount } from '@metamask/keyring-internal-api';
+
+import { USER_STORAGE_FEATURE_NAMES } from '../../../shared/storage-schema';
+import {
+  canPerformAccountSyncing,
+  getInternalAccountsList,
+  getUserStorageAccountsList,
+} from './sync-utils';
+import type { AccountSyncingConfig, AccountSyncingOptions } from './types';
+import {
+  doesInternalAccountHaveCorrectKeyringType,
+  isNameDefaultAccountName,
+  mapInternalAccountToUserStorageAccount,
+} from './utils';
+
+/**
+ * Saves an individual internal account to the user storage.
+ * @param internalAccount - The internal account to save
+ * @param config - parameters used for saving the internal account
+ * @param options - parameters used for saving the internal account
+ */
+export async function saveInternalAccountToUserStorage(
+  internalAccount: InternalAccount,
+  config: AccountSyncingConfig,
+  options: AccountSyncingOptions,
+): Promise<void> {
+  const { isAccountSyncingEnabled } = config;
+  const { getUserStorageControllerInstance } = options;
+
+  if (
+    !isAccountSyncingEnabled ||
+    !canPerformAccountSyncing(config, options) ||
+    !isEvmAccountType(internalAccount.type) ||
+    !doesInternalAccountHaveCorrectKeyringType(internalAccount)
+  ) {
+    return;
+  }
+
+  try {
+    // Map the internal account to the user storage account schema
+    const mappedAccount =
+      mapInternalAccountToUserStorageAccount(internalAccount);
+
+    await getUserStorageControllerInstance().performSetStorage(
```

Comment:
(not actionable, just a comment) Oooh... I just realised, that if we have the messenger, then you don't need the controller instance..

```ts
const messenger = options.getMessenger();
await messenger.call('UserStorageController:performSetStorage', ..., ....)
```

Hmm need to weigh pros/cons.

Pro:
- 1 less dependency to manage
- We technically don't need the entire controller instance, just methods, which this has.

Cons:
- I think tests (the messenger mock) may need to support this.

Reply:
Yeah no need to action this. Can be a future refactor if we want it.

But maybe we can store the instance as a variable so it can be reused (only if we call this alot).
```ts
const messenger = options.getMessenger()
const controller - options.getUserStorageControllerInstance();
```

Reply:
meh its fine

Reply:
This is exactly the type of thing the messenger is intended for! it's also intended to be very easy to mock (much easier than a controller).